26.2.0 GA release notes (#23205)
cockroachdb/docs: src/current/_includes/releases/v26.2/v26.2.0.md

Reviewer: TCP stuff LGTM - could provide a link to the session vars page for users to click

  <tr>
   <td>
    <p class="feature-summary">PostgreSQL-compatible fuzzystrmatch functions support</p>

>     <p class="feature-description">CockroachDB now supports PostgreSQL-compatible fuzzystrmatch built-in functions: <code>dmetaphone()</code>, <code>dmetaphone_alt()</code>, and <code>daitch_mokotoff()</code>. These functions are useful for fuzzy string matching based on phonetic similarity. <code>dmetaphone</code> and <code>dmetaphone_alt</code> return Double Metaphone phonetic codes for a string, and <code>daitch_mokotoff</code> returns an array of Daitch-Mokotoff soundex codes.</p>

Reviewer: suggest updating this to include more functions (just discussed with @rafiss on https://github.com/cockroachdb/docs/pull/23191#discussion_r3102591599):

- `soundex()`
- `difference()`
- `levenshtein()`
- `levenshtein_less_equal()`
- `metaphone()`
- `dmetaphone()`
- `dmetaphone_alt()`
- `daitch_mokotoff()`

with the longer list of functions i don't think it's necessary to also provide descriptions of what each of them do, probably best to link to https://www.cockroachlabs.com/docs/stable/functions-and-operators#fuzzy-string-matching-functions for the details

  </tr>
  <tr>
   <td>
    <p class="feature-summary">Schema lock enforcement</p>

>     <p class="feature-description">The cluster setting <code>sql.schema.auto_unlock.enabled</code> controls whether DDL operations automatically unlock <code>schema_locked</code> tables. When set to <code>false</code>, DDL statements on schema-locked tables are blocked unless manually unlocked. This allows users of Logical Data Replication (LDR) to enforce <code>schema_locked</code> as a hard lock preventing user-initiated DDL. The default is <code>true</code>.</p>

Reviewer: i'm not sure if it's worth adding this level of nuance in a feature highlight (cc @rafiss if you have opinions) - in [the docs PR](https://github.com/cockroachdb/docs/pull/23191/changes) we made it clear that it can't always auto-unlock

suggest clarifying slightly with replacement text something like the following (IFF we care about this level of nuance, which i think we will, since it could lead to some user/support interactions re: "why doesn't this auto-unlock?"):

```
<p class="feature-description">The cluster setting <code>sql.schema.auto_unlock.enabled</code> controls whether DDL operations attempt to automatically unlock and re-lock <code>schema_locked</code> tables. When set to <code>false</code>, DDL statements on schema-locked tables are blocked unless manually unlocked. This allows users of Logical Data
  Replication (LDR) to enforce <code>schema_locked</code> as a hard lock preventing user-initiated DDL. Some schema changes still require manual unlock even when this setting is <code>true</code>. The default is <code>true</code>.</p>
```

  </tr>
  <tr>
   <td>
    <p class="feature-summary">Hash-sharded indexes with prefix columns</p>

>     <p class="feature-description">Hash-sharded indexes now support computing the shard value from a subset of index key columns rather than all of them. This gives you finer control over how data is distributed across shards and significantly improves query performance when filtering on only a prefix of the indexed columns.</p>

Reviewer: this description can be updated to link to the docs for this which are live: https://www.cockroachlabs.com/docs/v26.2/hash-sharded-indexes#shard-columns

 <tbody>
  <tr>
   <td>
    <p class="feature-summary">Certificate-based authentication using X.509 Subject field</p>

>     <p class="feature-description">CockroachDB now supports mapping SQL user roles to distinguished name attributes in the Subject field of X.509 certificates, including <code>OU</code>, <code>UID</code>, and <code>CN</code>. The cluster setting <code>security.client_cert.san_required.enabled</code> optionally allows mapping to Subject Alternative Name (SAN) fields instead. This enables authentication using your existing Certificate Authority infrastructure without requiring <code>CommonName</code>-based restrictions.</p>

Reviewer 2: Should link to https://www.cockroachlabs.com/docs/v26.2/certificate-based-authentication-using-the-x509-subject-field.html#optional-enable-the-cluster-setting-to-map-users-to-subject-alternative-name-san-fields

   </td>
   <td class="icon-center">Preview</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
  </tr>
  <tr>
   <td>
    <p class="feature-summary">Post-quantum cryptography support</p>

>     <p class="feature-description">CockroachDB now supports post-quantum cryptographic algorithms for TLS connections. This applies to both client-to-node and inter-node communication.</p>

Reviewer 3: ```suggestion
    <p class="feature-description">CockroachDB now supports post-quantum cryptographic algorithms for TLS 1.3 connections. This applies to both client-to-node and inter-node communication.</p>
```

>     <p class="feature-description">CockroachDB now supports post-quantum cryptographic algorithms for TLS connections. This applies to both client-to-node and inter-node communication.</p>

Reviewer 3: Link to [https://www.cockroachlabs.com/docs/v26.2/security-reference/transport-layer-security#post-quantum-cryptography-support-in-cockroachdb](https://www.cockroachlabs.com/docs/v26.2/security-reference/transport-layer-security#post-quantum-cryptography-support-in-cockroachdb)

 <tbody>
  <tr>
   <td>
    <p class="feature-summary">Active Session History</p>

>     <p class="feature-description">Active Session History (ASH) tracks CPU usage, I/O activity, wait events, and contention for session activity including SQL statements and background jobs. Samples are captured at regular intervals, enabling faster diagnosis of performance bottlenecks by correlating session activity with resource consumption.</p>

Reviewer 3: Add link to [https://www.cockroachlabs.com/docs/v26.2/active-session-history](https://www.cockroachlabs.com/docs/v26.2/active-session-history)

   </td>
   <td class="icon-center">GA</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
  </tr>
 </tbody>
</table>

</div>

<a id="v26-2-0-backward-incompatible-changes"></a>

### Features that require upgrade finalization

{% include releases/v26.2/upgrade-finalization.md %}

### Backward-incompatible changes

{% include releases/v26.2/backward-incompatible.md %}

<h4 id="v26-2-0-cluster-settings">Key cluster setting changes</h4>

{% include releases/v26.2/cluster-setting-changes.md %}

<h4 id="v26-2-0-deprecations">Deprecations</h4>

{% include releases/v26.2/deprecations.md %}

<h3 id="v26-2-0-security-updates">Security updates</h3>

- LDAP authentication for the DB Console now supports automatic user provisioning. When the cluster setting `security.provisioning.ldap.enabled` is set to true, users who authenticate successfully via LDAP will be automatically created in CockroachDB if they do not already exist. [#163199][#163199]

> - The new cluster setting `security.client_cert.san_required.enabled` enables Subject Alternative Name (SAN) based authentication for client certificates. When enabled, CockroachDB validates client identities using SAN attributes (URIs, DNS names, or IP addresses) from X.509 certificates instead of or in addition to the certificate's Common Name field.

Reviewer 2: Link to https://www.cockroachlabs.com/docs/v26.2/certificate-based-authentication-using-the-x509-subject-field.html#optional-enable-the-cluster-setting-to-map-users-to-subject-alternative-name-san-fields

    This authentication method works across both SQL client connections and internal RPC communication between cluster nodes, ensuring consistent identity verification throughout the system. Organizations using modern certificate management systems and service identity frameworks can now leverage their existing infrastructure for database authentication without requiring certificate reissuance or CN-based naming conventions. [#162583][#162583]
- When the `security.provisioning.ldap.enabled` cluster setting is enabled, LDAP-authenticated DB Console logins now update the `estimated_last_login_time` column in the `system.users` table. [#163400][#163400]
- When the `security.provisioning.oidc.enabled` cluster setting is enabled, OIDC-authenticated DB Console logins now populate the `estimated_last_login_time` column in `system.users`, allowing administrators to track when OIDC users last accessed the DB Console. [#164129][#164129]
- Removed an overly restrictive TLS curve preference that limited FIPS mode to P-256. CockroachDB now uses Go's native FIPS curve selection, improving interoperability with clients that prefer other FIPS curves. [#166793][#166793]

> <h3 id="v26-2-0-enterprise-edition-changes">Enterprise edition changes</h3>

Reviewer 2: We need a better way of calling out "these are changes that aren't available on trial/evaluation licenses", if that's important to call out, than leaning on the word "enterprise". If not for this quarter, then definitely next.

Author: Agree; just added a callout for now that links to the Licensing FAQs.

- `ALTER TABLE ... SET LOCALITY` is now fully executed using the declarative schema changer, improving reliability and consistency with other schema change operations. [#161763][#161763]
- Setting `skip_unique_checks = true` on an index now emits a notice warning that unique constraint enforcement is bypassed, with a pointer to the `INSPECT` documentation. [#167405][#167405]

<h3 id="v26-2-0-operational-changes">Operational changes</h3>

- Changefeeds now support the `partition_alg` option for specifying a Kafka partitioning algorithm. Currently `fnv-1a` (default) and `murmur2` are supported. The option is only valid on Kafka v2 sinks. This is protected by the cluster setting `changefeed.partition_alg.enabled`. An example usage: `SET CLUSTER SETTING changefeed.partition_alg.enabled=true; CREATE CHANGEFEED ... INTO 'kafka://...' WITH partition_alg='murmur2';`. Note that if a changefeed is created using the `murmur2` algorithm, and then the cluster setting is disabled, the changefeed will continue using the `murmur2` algorithm unless the changefeed is altered to use a different `partition_alg`. [#161265][#161265]

> - Added the `server.sql_tcp_user.timeout` cluster setting, which specifies the maximum amount of time transmitted data can remain unacknowledged before the underlying TCP connection is forcefully closed. This setting is enabled by default with a value of 30 seconds and is supported on Linux and macOS (Darwin). [#164037][#164037]

Reviewer 2: Was the previous behavior that TCP connections were not closed at all, and now we're enforcing a 30 second default? If so this smells to me like a change that could have unexpected impact in edge cases, aka a potential breaking change.

Reviewer 2: Any new settings that modify existing behavior and default to an "opt out" state set off my alarm bells

Author: Thanks for calling this out. I'm moving it to backward-incompatible.

Author: Well, based on the original PR this is about "detecting and closing dead connections faster than the default TCP keep-alive mechanism in some scenarios". But it sounds like it could potentially break things.

 <tbody>
  <tr>
   <td>
    <p class="feature-summary">Leader leases</p>

>     <p class="feature-description">Leader leases are now generally available. This feature maintains more stable leadership across nodes by reducing unnecessary lease transfers, resulting in more consistent query response times and fewer latency spikes.</p>

Reviewer: 'Leader Leases' can be a link to https://www.cockroachlabs.com/docs/v26.2/architecture/replication-layer.html#leader-leases

   </td>
   <td class="icon-center">GA</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
  </tr>
  <tr>
   <td>
    <p class="feature-summary">Buffered writes</p>

>     <p class="feature-description">Buffered writes are now generally available. This feature improves throughput and reducing tail latency under heavy write workloads by batching writes efficiently before flushing to disk.</p>

Reviewer: 'buffered writes' can be a link to the docs at https://www.cockroachlabs.com/docs/v26.2/architecture/transaction-layer#buffered-writes

(PS the PR to update them as being GA is in SME review now, should merge for release)

- Improved changefeed checkpointing performance when changefeeds are lagging. Previously, checkpoint updates could be redundantly applied multiple times per checkpoint operation. [#162546][#162546]
- Queries that have comparison expressions with the `levenshtein` built-in are now up to 30% faster. [#160394][#160394]
- Fixed a performance regression in `pg_catalog.pg_roles` and `pg_catalog.pg_authid` by avoiding privilege lookups for each row in the table. [#160121][#160121]
- Significantly reduced WAL write latency when using encryption at rest by properly recycling WAL files instead of deleting and recreating them. [#160784][#160784]
- Optimized the logic that applies zone config constraints so it no longer fetches all descriptors in the cluster during background constraint reconciliation. [#160966][#160966]
- Various background tasks and jobs now more actively yield to foreground work when that work is waiting to run. [#159205][#159205]
- Statement executions using canary stats will no longer use cached plans, which prevents cache thrashing but causes a slight increase in planning time over statement executions using stable stats. [#167503][#167503]

> ### Known limitations

Reviewer 3: These are the known limitations for Active Session History, so I think they belong here?

- ASH is not recommended for nodes with 64 or more vCPUs, due to degraded performance on those nodes. [#168289](https://github.com/cockroachdb/cockroach/issues/168289)
- On Basic and Standard CockroachDB {{ site.data.products.cloud }} clusters, ASH samples only cover work running on the [SQL]({% link {{ page.version.version }}/architecture/sql-layer.md %}) pod. KV-level work ([storage]({% link {{ page.version.version }}/architecture/storage-layer.md %}) I/O, [lock waits]({% link {{ page.version.version }}/troubleshoot-lock-contention.md %}), [replication]({% link {{ page.version.version }}/architecture/replication-layer.md %}), etc.) is not visible in ASH samples.
- KV work triggered during [COMMIT]({% link {{ page.version.version }}/commit-transaction.md %}) (for example, [intent resolution]({% link {{ page.version.version }}/architecture/transaction-layer.md %}), [Raft]({% link {{ page.version.version }}/architecture/replication-layer.md %}#raft) proposals deferred from earlier statements in an [explicit transaction]({% link {{ page.version.version }}/begin-transaction.md %})) is attributed to the last [statement's fingerprint]({% link {{ page.version.version }}/ui-statements-page.md %}), not the statement that originally caused the work. [#165864](https://github.com/cockroachdb/cockroach/issues/165864)`

- Queries that have comparison expressions with the `levenshtein` built-in are now up to 30% faster. [#160394][#160394]
- Fixed a performance regression in `pg_catalog.pg_roles` and `pg_catalog.pg_authid` by avoiding privilege lookups for each row in the table. [#160121][#160121]
- Significantly reduced WAL write latency when using encryption at rest by properly recycling WAL files instead of deleting and recreating them. [#160784][#160784]
- Optimized the logic that applies zone config constraints so it no longer fetches all descriptors in the cluster during background constraint reconciliation. [#160966][#160966]
- Various background tasks and jobs now more actively yield to foreground work when that work is waiting to run. [#159205][#159205]
- Statement executions using canary stats will no longer use cached plans, which prevents cache thrashing but causes a slight increase in planning time over statement executions using stable stats. [#167503][#167503]

### Known limitations

This section describes newly identified limitations in CockroachDB v26.2.

> #### Hint injections

Reviewer 3: ```suggestion
#### Statement hints
```
Also the two "hint injection" KLs should be replaced with an include to _includes/v26.2/known-limitations/statement-hints-limitations.md

That include file will be present once [this PR](https://github.com/cockroachdb/docs/pull/23220) gets approved and merged into main (and then once main gets merged into your PR)

    <p class="feature-description"><a href="{% link v26.2/architecture/transaction-layer.md %}#buffered-writes">Buffered writes</a> are now generally available. This feature improves throughput and reducing tail latency under heavy write workloads by batching writes efficiently before flushing to disk.</p>
   </td>
   <td class="icon-center">GA</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
   <td class="icon-center">{% include icon-yes.html %}</td>
  </tr>
 </tbody>
</table>

</div>

<a id="v26-2-0-backward-incompatible-changes"></a>

### Features that require upgrade finalization

{% include releases/v26.2/upgrade-finalization.md %}

### Backward-incompatible changes

{% include releases/v26.2/backward-incompatible.md %}

<h4 id="v26-2-0-cluster-settings">Key cluster setting changes</h4>

{% include releases/v26.2/cluster-setting-changes.md %}

<h4 id="v26-2-0-deprecations">Deprecations</h4>

{% include releases/v26.2/deprecations.md %}

<h3 id="v26-2-0-security-updates">Security updates</h3>

> - LDAP authentication for the DB Console now supports automatic user provisioning. When the cluster setting `security.provisioning.ldap.enabled` is set to true, users who authenticate successfully via LDAP will be automatically created in CockroachDB if they do not already exist. [#163199][#163199]

Reviewer 4: ```suggestion
- LDAP automatic user provisioning now extends to the DB Console in addition to SQL client connections. When `security.provisioning.ldap.enabled` is set to true, users authenticating via LDAP are automatically created in CockroachDB if they do not already exist. For more information, refer to [LDAP Authentication]({% link {{ page.version.version }}/ldap-authentication.md %}#option-1-automatic-user-provisioning-recommended). [#163199][#163199]
```

@taroface FYI the docs link here already exists, but those docs are about to be updated today when https://github.com/cockroachdb/docs/pull/23213 merges. (My only remaining open docs PR for the release. Just needs tech writer review.)
